Enabling BuildCheck causes  error MSB4166: Child node "1" exited prematurely

### Issue Description

Enabling BuildCheck on runtime and other repos caused

```
 error MSB4166: Child node "1" exited prematurely. Shutting down. Diagnostic information may be found in files in "C:\Users\cloudtest\AppData\Local\Temp\MSBuildTemp\" and will be named MSBuild_*.failure.txt. This location can be changed by setting the MSBUILDDEBUGPATH environment variable to a different directory.Microsoft.Build.Framework.InternalErrorException: MSB0001: Internal MSBuild Error: ContextID 415 should have been in the ID-to-project file mapping but wasn't!
   at Microsoft.Build.Shared.ErrorUtilities.ThrowInternalError(String message, Object[] args)
   at Microsoft.Build.BackEnd.Logging.LoggingService.GetAndVerifyProjectFileFromContext(BuildEventContext context)
   at Microsoft.Build.BackEnd.Logging.LoggingService.LogBuildEvent(BuildEventArgs buildEvent)
   at Microsoft.Build.Experimental.BuildCheck.Infrastructure.BuildCheckManagerProvider.BuildCheckManager.StartProjectRequest(ICheckContext checkContext, String projectFullPath)
   at Microsoft.Build.Experimental.BuildCheck.Infrastructure.BuildCheckBuildEventHandler.HandleProjectStartedRequest(ProjectStartedEventArgs eventArgs)
   at Microsoft.Build.Experimental.BuildCheck.Infrastructure.BuildCheckConnectorLogger.EventSource_AnyEventRaised(Object sender, BuildEventArgs e)
   at Microsoft.Build.BackEnd.Logging.EventSourceSink.RaiseAnyEvent(BuildEventArgs buildEvent)
   at Microsoft.Build.BackEnd.Logging.EventSourceSink.<Consume>b__81_9(ProjectStartedEventArgs args)
   at Microsoft.Build.BackEnd.Logging.EventSourceSink.Consume(BuildEventArgs buildEvent)
   at Microsoft.Build.BackEnd.Logging.EventSourceSink.RaiseAnyEvent(BuildEventArgs buildEvent)
   at Microsoft.Build.BackEnd.Logging.EventSourceSink.<Consume>b__81_9(ProjectStartedEventArgs args)
   at Microsoft.Build.BackEnd.Logging.EventSourceSink.Consume(BuildEventArgs buildEvent)
   at Microsoft.Build.BackEnd.Logging.LoggingService.RouteBuildEvent(BuildEventArgs eventArg)
   at Microsoft.Build.BackEnd.Logging.LoggingService.RouteBuildEvent(Object loggingEvent)
   at Microsoft.Build.BackEnd.Logging.LoggingService.LoggingEventProcessor(Object loggingEvent)
```

or

```
##[error]Child node "1" exited prematurely. Shutting down. Diagnostic information may be found in files in "/tmp/MSBuildTempcloudtest/" and will be named MSBuild_*.failure.txt. This location can be changed by setting the MSBUILDDEBUGPATH environment variable to a different directory.Microsoft.Build.Framework.InternalErrorException: MSB0001: Internal MSBuild Error: ContextID 61 should have been in the ID-to-project file mapping but wasn't!
   at Microsoft.Build.Shared.ErrorUtilities.ThrowInternalError(String message, Object[] args)
   at Microsoft.Build.BackEnd.Logging.LoggingService.GetAndVerifyProjectFileFromContext(BuildEventContext context)
   at Microsoft.Build.BackEnd.Logging.LoggingService.LogBuildEvent(BuildEventArgs buildEvent)
   at Microsoft.Build.Experimental.BuildCheck.Infrastructure.BuildCheckManagerProvider.BuildCheckManager.Microsoft.Build.Experimental.BuildCheck.Infrastructure.IResultReporter.ReportResult(BuildEventArgs eventArgs, ICheckContext checkContext)
   at Microsoft.Build.Experimental.BuildCheck.Infrastructure.CheckWrapper.ReportResult(BuildCheckResult result, ICheckContext checkContext, CheckConfigurationEffective config)
   at Microsoft.Build.Experimental.BuildCheck.Infrastructure.BuildEventsProcessor.ReportResult(CheckWrapper checkWrapper, ICheckContext checkContext, CheckConfigurationEffective[] configPerRule, BuildCheckResult result)
   at Microsoft.Build.Experimental.BuildCheck.Checks.PropertiesUsageCheck.DoneWithProject(BuildCheckDataContext`1 context)
   at Microsoft.Build.Experimental.BuildCheck.Infrastructure.BuildCheckCentralContext.<>c__DisplayClass25_0`1.<RegisterAction>g__WrappedHandler|0(BuildCheckDataContext`1 context)
   at Microsoft.Build.Experimental.BuildCheck.Infrastructure.BuildCheckCentralContext.RunRegisteredActions[T](List`1 registeredCallbacks, T checkData, ICheckContext checkContext, Action`4 resultHandler)
   at Microsoft.Build.Experimental.BuildCheck.Infrastructure.BuildEventsProcessor.ProcessProjectDone(ICheckContext checkContext, String projectFullPath)
   at Microsoft.Build.Experimental.BuildCheck.Infrastructure.BuildCheckBuildEventHandler.HandleBuildEvent(BuildEventArgs e)
   at Microsoft.Build.BackEnd.Logging.EventSourceSink.RaiseAnyEvent(Object sender, BuildEventArgs buildEvent)
   at Microsoft.Build.BackEnd.Logging.EventSourceSink.RaiseAnyEvent(Object sender, BuildEventArgs buildEvent)
   at Microsoft.Build.BackEnd.Logging.LoggingService.RouteBuildEvent(BuildEventArgs eventArg)
   at Microsoft.Build.BackEnd.Logging.LoggingService.RouteBuildEvent(Object loggingEvent)
   at Microsoft.Build.BackEnd.Logging.LoggingService.LoggingEventProcessor(Object loggingEvent)

```

It fails randomly but for sure caused by the feature.

### Steps to Reproduce

navigate to the repo and observe the issue https://github.com/dotnet/extensions/pull/5810/checks?check_run_id=35940423010 or check the binlog attached (change .zip -> .binlog)

[Build_runtime_windows.zip](https://github.com/user-attachments/files/18522252/Build_runtime_windows.zip)

### Expected Behavior

BuildCheck runs smootly.

### Actual Behavior

enabling the feature causes build crash.

### Analysis

_No response_

### Versions & Configurations

_No response_

--- a/src/Build/BackEnd/Components/Logging/LoggingService.cs
+++ b/src/Build/BackEnd/Components/Logging/LoggingService.cs
@@ -73,6 +73,11 @@ namespace Microsoft.Build.BackEnd.Logging
     internal partial class LoggingService : ILoggingService, INodePacketHandler
     {
         /// <summary>
+        /// Gets or sets a value if BuildCheck is enabled. The presence of this flag influences the logging logic.
+        /// </summary>
+        private bool _buildCheckEnabled;
+
+        /// <summary>
         /// The default maximum size for the logging event queue.
         /// </summary>
         private const uint DefaultQueueCapacity = 200000;
@@ -474,11 +479,6 @@ namespace Microsoft.Build.BackEnd.Logging
         public bool Question { get; set; }
 
         /// <summary>
-        /// Gets or sets a value if BuildCheck is enabled. The presence of this flag influences the logging logic.
-        /// </summary>
-        internal bool BuildCheckEnabled { get; set; }
-
-        /// <summary>
         /// The list of descriptions which describe how to create forwarding loggers on a node.
         /// This is used by the node provider to get a list of registered descriptions so that
         /// they can be transmitted to child nodes.
@@ -877,7 +877,7 @@ namespace Microsoft.Build.BackEnd.Logging
 
                 _buildEngineDataRouter = (buildComponentHost.GetComponent(BuildComponentType.BuildCheckManagerProvider) as IBuildCheckManagerProvider)?.BuildEngineDataRouter;
 
-                BuildCheckEnabled = buildComponentHost.BuildParameters!.IsBuildCheckEnabled;
+                _buildCheckEnabled = buildComponentHost.BuildParameters!.IsBuildCheckEnabled;
             }
         }
 

--- a/src/Build/BackEnd/Components/Logging/LoggingService.cs
+++ b/src/Build/BackEnd/Components/Logging/LoggingService.cs
@@ -1,4 +1,4 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
+// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
@@ -877,7 +877,7 @@ namespace Microsoft.Build.BackEnd.Logging
 
                 _buildEngineDataRouter = (buildComponentHost.GetComponent(BuildComponentType.BuildCheckManagerProvider) as IBuildCheckManagerProvider)?.BuildEngineDataRouter;
 
-                _buildCheckEnabled = buildComponentHost.BuildParameters!.IsBuildCheckEnabled;
+                _buildCheckEnabled = buildComponentHost.BuildParameters.IsBuildCheckEnabled;
             }
         }
 

--- a/src/Build/BackEnd/Components/Logging/LoggingService.cs
+++ b/src/Build/BackEnd/Components/Logging/LoggingService.cs
@@ -474,6 +474,11 @@ namespace Microsoft.Build.BackEnd.Logging
         public bool Question { get; set; }
 
         /// <summary>
+        /// Gets or sets a value if BuildCheck is enabled. The presence of this flag influences the logging logic.
+        /// </summary>
+        internal bool BuildCheckEnabled { get; set; }
+
+        /// <summary>
         /// The list of descriptions which describe how to create forwarding loggers on a node.
         /// This is used by the node provider to get a list of registered descriptions so that
         /// they can be transmitted to child nodes.
@@ -871,6 +876,8 @@ namespace Microsoft.Build.BackEnd.Logging
                 _serviceState = LoggingServiceState.Initialized;
 
                 _buildEngineDataRouter = (buildComponentHost.GetComponent(BuildComponentType.BuildCheckManagerProvider) as IBuildCheckManagerProvider)?.BuildEngineDataRouter;
+
+                BuildCheckEnabled = buildComponentHost.BuildParameters!.IsBuildCheckEnabled;
             }
         }
 

--- a/src/Build/BackEnd/Components/Logging/LoggingServiceLogMethods.cs
+++ b/src/Build/BackEnd/Components/Logging/LoggingServiceLogMethods.cs
@@ -617,7 +617,7 @@ namespace Microsoft.Build.BackEnd.Logging
 
             // BuildCheck can still emit some LogBuildEvent(s) after ProjectFinishedEventArgs was reported.
             // Due to GetAndVerifyProjectFileFromContext validation, these checks break the build.
-            if (!BuildCheckEnabled)
+            if (!_buildCheckEnabled)
             {
                 // PERF: Not using VerifyThrow to avoid boxing of projectBuildEventContext.ProjectContextId in the non-error case.
                 if (!_projectFileMap.TryRemove(projectBuildEventContext.ProjectContextId, out _))

--- a/src/Build/BackEnd/Components/Logging/LoggingServiceLogMethods.cs
+++ b/src/Build/BackEnd/Components/Logging/LoggingServiceLogMethods.cs
@@ -615,10 +615,15 @@ namespace Microsoft.Build.BackEnd.Logging
             buildEvent.BuildEventContext = projectBuildEventContext;
             ProcessLoggingEvent(buildEvent);
 
-            // PERF: Not using VerifyThrow to avoid boxing of projectBuildEventContext.ProjectContextId in the non-error case.
-            if (!_projectFileMap.TryRemove(projectBuildEventContext.ProjectContextId, out _))
+            // BuildCheck can still emit some LogBuildEvent(s) after ProjectFinishedEventArgs was reported.
+            // Due to GetAndVerifyProjectFileFromContext validation, these checks break the build.
+            if (!BuildCheckEnabled)
             {
-                ErrorUtilities.ThrowInternalError("ContextID {0} for project {1} should be in the ID-to-file mapping!", projectBuildEventContext.ProjectContextId, projectFile);
+                // PERF: Not using VerifyThrow to avoid boxing of projectBuildEventContext.ProjectContextId in the non-error case.
+                if (!_projectFileMap.TryRemove(projectBuildEventContext.ProjectContextId, out _))
+                {
+                    ErrorUtilities.ThrowInternalError("ContextID {0} for project {1} should be in the ID-to-file mapping!", projectBuildEventContext.ProjectContextId, projectFile);
+                }
             }
         }
 

